tools/clang/include/clang/SPIRV/SpirvInstruction.h

```diff
@@ -1302,6 +1302,22 @@ class SpirvConstantNull : public SpirvConstant {
   bool operator==(const SpirvConstantNull &that) const;
 };
 
+class SpirvUndef : public SpirvConstant {
```

Comment:
This is awkward. None of the extra functions provided by SpirvConstant even make sense. For example, what happens if you call `setLiteral` on an undef? Is that meaning full?

However, because an undef could be a constituent of an OpCompositeComposite, you want to have parent type for all constants and undefs.

I don't think we should make major changes, but we should add comments to explain why an Undef is being treated as a constant, and have some type of error checking somewhere that an undef is not suppose to be a literal.

Reply:
Yes, I agree it's weird. I made the change to actually inherit from SpirvInstr, and it was fine.
The main reason is: we never generate OpUndef in an OpComposite. The only case it seems we can generate such instruction is with those dead branches, otherwise it's just an variable we haven't initialized. So it seems fine to split SpirvConstant and SpirvUndef.